File: clojure/plugins/voxgig/sekreto/plugins/hashicorp.clj
;; HashiCorp Vault, as a voxgig/plugin definition.
;;
;; A plugin rather than a built-in kind because it opens a socket: a chain
;; that names `hashicorp` must be handed this definition by the calling
;; project (docs/design/plugin-providers.md).
;;
;; A port of typescript/plugins/hashicorp.ts, which is canonical.

(ns voxgig.sekreto.plugins.hashicorp
  (:require [clojure.string :as string]
            [voxgig.sekreto.core :as core]
            [voxgig.sekreto.addr :refer [checkaddr]]
            [voxgig.sekreto.json :as json]
            [voxgig.sekreto.provider :as provider]
            [voxgig.sekreto.providers :as providers]
            [voxgig.sekreto.plugins.httpjson :as http]))

(defrecord Hashicorp [addr token mount kv vaultnamespace auth state]
  provider/Provider

  ;; HashiCorp Vault.
  ;;
  ;; KV v2 (the default): `api.token` reads `{addr}/v1/{mount}/data/api` and
  ;; takes the `token` field of `data.data`. KV v1 (`kv: 1`) reads
  ;; `{addr}/v1/{mount}/api` and takes the field of `data`. A 404 means "not
  ;; here" - a miss - so a vault can sit in a chain with fallbacks.
  ;;
  ;; A Vault Enterprise namespace rides the X-Vault-Namespace header, on
  ;; logins as well as reads.
  ;;
  ;; Instead of being handed a token, the provider can log in: Kubernetes
  ;; auth (the pod's service-account JWT, from its conventional path) or
  ;; AppRole. A failed login is an error, never a miss - it means this store
  ;; could not answer at all.
  (lookup [_ name]
    (checkaddr addr)

    (let [baseheaders (if (core/notempty vaultnamespace)
                        {"X-Vault-Namespace" vaultnamespace}
                        {})

          login (fn []
                  (when (nil? auth)
                    (throw (core/sekretoerror "sekreto: hashicorp: no token and no auth method")))

                  (let [method (:method auth)
                        url (str (http/trimslash addr) "/v1/auth/"
                                 (http/firstof (:mount auth) method) "/login")
                        body (case method
                               "kubernetes"
                               (let [file (or (core/notempty (:jwtfile auth))
                                              "/var/run/secrets/kubernetes.io/serviceaccount/token")
                                     jwt (or (core/notempty (:jwt auth))
                                             (core/notempty
                                              (some-> (providers/readfile (providers/path-of file) "hashicorp") string/trim))
                                             (throw (core/sekretoerror
                                                     (str "sekreto: hashicorp: cannot read jwt file "
                                                          file))))]
                                 (json/omap [["role" (or (:role auth) "")] ["jwt" jwt]]))

                               "approle"
                               (json/omap [["role_id" (or (:roleid auth) "")]
                                           ["secret_id" (or (:secretid auth) "")]])

                               (throw (core/sekretoerror
                                       (str "sekreto: hashicorp: unknown auth method: " method))))

                        res (http/fetchjson "POST" url baseheaders (json/stringify body))
                        got (json/astext (json/dig (:body res) "auth" "client_token"))]

                    (when (or (not= 200 (:status res)) (nil? (core/notempty got)))
                      (throw (core/sekretoerror
                              (str "sekreto: hashicorp login failed: " (:status res) ": " url))))

                    [got (http/renewtime (json/dig (:body res) "auth" "lease_duration"))]))

          live (http/withtoken state login)
          ref (core/vaultref name)
          base (str (http/trimslash addr) "/v1/" mount)
          url (if (= 1 kv)
                (str base "/" (:path ref))
                (str base "/data/" (:path ref)))

          res (http/fetchjson "GET" url (assoc baseheaders "X-Vault-Token" (or live "")))]

      (cond
        (= 404 (:status res)) nil

        (not= 200 (:status res))
        (throw (core/sekretoerror (str "sekreto: hashicorp error: " (:status res) ": " url)))

        :else
        (let [data (if (= 1 kv)
                     (json/dig (:body res) "data")
                     (json/dig (:body res) "data" "data"))]
          (json/astext (json/dig data (:field ref)))))))

  (describe [_] (str "hashicorp:" addr "/" mount)))

(def hashicorp
  "The `hashicorp` provider kind."
  (providers/providerplugin
   "hashicorp"
   (fn [{:keys [addr token mount kv vaultnamespace auth]}]
     (let [usekv (or kv 2)]
       ;; A version typo like kv: 3 must not quietly behave as v2 and turn
       ;; its 404s into misses; there is nothing safe to assume it meant.
       (when (and (not= 1 usekv) (not= 2 usekv))
         (throw (core/sekretoerror (str "sekreto: hashicorp: unsupported kv version: " usekv))))
       (->Hashicorp (or addr "") (core/notempty token) (or (core/notempty mount) "secret")
                    usekv vaultnamespace auth
                    (atom {:token (core/notempty token) :renewat Long/MAX_VALUE}))))))

File: lua/src/sekreto/plugins/hashicorp.lua
-- The hashicorp plugin: HashiCorp Vault, over HTTPS. Needs a socket, so
-- it is not a built-in kind - a chain that never names it never requires
-- this file (docs/design/plugin-providers.md).
--
-- A port of typescript/plugins/hashicorp.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')

local fail = err.fail
local nonempty = providers.nonempty
local trimslash = providers.trimslash
local readfile = providers.readfile
local providerplugin = providers.providerplugin
local first = support.first
local nowms = support.nowms
local renewtime = support.renewtime
local NEVER = support.NEVER
local fetchjson = httpjson.fetchjson

local M = {}

--- HashiCorp Vault.
---
--- KV v2 (the default): `api.token` reads `{addr}/v1/{mount}/data/api`
--- and takes the `token` field of `data.data`. KV v1 reads
--- `{addr}/v1/{mount}/api` and takes the field of `data`. A 404 means
--- "not here" - a miss - so a vault can sit in a chain with fallbacks.
---
--- A Vault Enterprise namespace rides the X-Vault-Namespace header, on
--- logins as well as reads.
---
--- Instead of being handed a token, the provider can log in: Kubernetes
--- auth (the pod's service-account JWT, from its conventional path) or
--- AppRole. A failed login is an error, never a miss - it means this
--- store could not answer at all.
local function hashicorp(useaddr, token, mount, kv, vaultnamespace, auth)
  local address = useaddr or ''
  local usemount = nonempty(mount) and mount or 'secret'
  local usekv = kv or 2

  -- A version typo like kv: 3 must not quietly behave as v2 and turn its
  -- 404s into misses; there is nothing safe to assume it meant.
  if 1 ~= usekv and 2 ~= usekv then
    fail('sekreto: hashicorp: unsupported kv version: ' .. json.numstr(usekv))
  end

  local livetoken = nonempty(token) and token or nil
  local renewat = NEVER

  local function baseheaders()
    local out = {}
    if nonempty(vaultnamespace) then
      out[#out + 1] = { 'X-Vault-Namespace', vaultnamespace }
    end
    return out
  end

  local function login()
    if nil == auth then
      fail('sekreto: hashicorp: no token and no auth method')
    end

    local authmount = first(auth.mount, auth.method)
    local url = trimslash(address) .. '/v1/auth/' .. authmount .. '/login'
    local body

    if 'kubernetes' == auth.method then
      local jwt = auth.jwt

      if nil == jwt then
        local file = auth.jwtfile or
          '/var/run/secrets/kubernetes.io/serviceaccount/token'
        local text = readfile(file)
        if nil == text then
          fail('sekreto: hashicorp: cannot read jwt file ' .. file)
        end
        jwt = name.trim(text)
      end

      body = json.obj({ { 'role', auth.role or '' }, { 'jwt', jwt } })
    elseif 'approle' == auth.method then
      body = json.obj({
        { 'role_id', auth.roleid or '' },
        { 'secret_id', auth.secretid or '' },
      })
    else
      fail('sekreto: hashicorp: unknown auth method: ' .. tostring(auth.method))
    end

    local res = fetchjson('POST', url, baseheaders(), json.stringify(body))
    local got = json.text(json.dig(res.body, 'auth', 'client_token'))

    if 200 ~= res.status or nil == got or '' == got then
      fail('sekreto: hashicorp login failed: ' .. res.status .. ': ' .. url)
    end

    renewat = renewtime(json.dig(res.body, 'auth', 'lease_duration'))

    return got
  end

  return {
    lookup = function(secret)
      addr.checkaddr(address)

      if nil == livetoken or nowms() >= renewat then
        livetoken = login()
      end

      local ref = name.vaultref(secret)
      local base = trimslash(address) .. '/v1/' .. usemount
      local url = (1 == usekv) and (base .. '/' .. ref.path)
        or (base .. '/data/' .. ref.path)

      local headers = baseheaders()
      headers[#headers + 1] = { 'X-Vault-Token', livetoken or '' }

      local res = fetchjson('GET', url, headers)

      if 404 == res.status then
        return nil
      end

      if 200 ~= res.status then
        fail('sekreto: hashicorp error: ' .. res.status .. ': ' .. url)
      end

      local data = (1 == usekv) and json.dig(res.body, 'data')
        or json.dig(res.body, 'data', 'data')

      return json.text(json.dig(data, ref.field))
    end,
    describe = function()
      return 'hashicorp:' .. address .. '/' .. usemount
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.hashicorp = providerplugin('hashicorp', function(spec)
  return hashicorp(spec.addr, spec.token, spec.mount, spec.kv,
    spec.vaultnamespace, spec.auth)
end)

return M
